docsource/ops.rst

```diff
@@ -1,6 +1,228 @@
+        "version_type": "external"
+      },
+      "script": {
+        "lang": "painless",
+        "inline": "if (ctx._source.foo == 'bar') {ctx._version++; ctx._source.remove('foo')}"
+      }
+    }
+
+6. Now, reindex the entire data from current index to new index. Alternately, use a query to reindex only a subset of the current index.
+
+.. code-block:: bash
+
+    POST _reindex
+    {
+      "source": {
+        "index": "kbase.genome_1"
+      },
+      "dest": {
+        "index": "kbase.genome_1_r1"
+      }
+    }
+
+        OR
+
+    POST _reindex
+    {
+      "source": {
+        "index": "kbase.genome_1",
+        "query": {
+          ...
+        }
+      },
+      "dest": {
+        "index": "kbase.genome_1_r1"
+      }
+    }
+
+7. Run a checksum on the new index to make sure the numbers line up with the numbers of the current index.
+
+.. code-block:: bash
+
+    GET /kbase.genome_1_r1/_stats/docs,store
+
+8. Run a query to specifically check the change that was applied.
+
+.. code-block:: bash
+
+    GET kbase.genome_1_r1/_search
+
+   OR
+
+    GET kbase.genome_1_r1/_search
+    {
+     "query": {
+       "match": {
+         "FIELD": "VALUE"
+       }
+     }
+    }
+
+   OR
+
+    https://www.elastic.co/guide/en/elasticsearch/reference/5.5/search-request-body.html
+
+9. If the new index looks good, update index alias and delete current index.
+
+.. note::
+
+    If you want the current index to linger for a day or two to serve a rollback option, reindex the current index into another new index called kbase.genome_1_backup and then delete the current index. This is one of two ways of renaming an index in ElasticSearch. The other way is to use the snapshot API.
+
+.. code-block:: bash
+
+    POST _aliases
+    {
+     "actions": [
+     {
+       "add": {
+         "index": "kbase.genome_1_r1",
+         "alias": "kbase.genome"
+         }
+       },
+       {
+         "remove": {
+         "index": "kbase.genome_1",
+         "alias": "kbase.genome"
+       }
+     }
+     ]
+    }
+
+    DELETE kbase.genome_1
+
+10. List all available indexes for the genome alias and all available genome indexes to ensure consistency across the alias map. Verify that all genome indexes that are present (except for backups) are referenced by the alias. Also verify that the alias does not contain an index reference for which no index exists.
+
+.. code-block:: bash
+
+    GET /_cat/aliases/kbase.genome
+
+    GET /_cat/indices/kbase.genome_*
+
+11. If the change involved in the reindexing operation also requires a corresponding search type spec change (located in resources/types/genome.yml for example), then this change must be applied.
+
+12. Change mapping version from "1" to "_r1" in the resources/types/genome.yml search type spec and add a comment (for future reference) that describes the change that took place in the r1 reindexing operation.
+
+13. Restart trickle updates from the current time noted in step 1.
```

Comment:
Given these instructions I don't understand how reindexing would work when data needs to be fetched from the data source (say a new field is added to the spec).

Reply:
a) Change the type mapping, b) change the spec, generate new events for some time range in which the new field was added and run trickle updates (updates that take place as events trickle in). So far, I have defined a manual process with the steps a and b. Once we have a CLI for creating an index from a spec, it will be slightly more automated (with b being done before a) but the general idea still remains the same. This also assumes that we modify Search to overwrite existing data.

Reply:
> generate new events for some time range in which the new field was added

I'm not seeing this in the instructions unless I'm missing something? It's always a reindex from the existing ES data AFAICT.

Reply:
Ah, good point. Added.